sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java

```diff
@@ -164,25 +165,58 @@ public void upload(InputStream data, long length, boolean overwrite) {
      * @param requestConditions {@link BlobRequestConditions}
      * @param timeout An optional timeout value beyond which a {@link RuntimeException} will be raised.
      * @param context Additional context that is passed through the Http pipeline during the service call.
+     * @deprecated
      */
+    @Deprecated
     public void uploadWithResponse(InputStream data, long length, ParallelTransferOptions parallelTransferOptions,
         BlobHttpHeaders headers, Map<String, String> metadata, AccessTier tier, BlobRequestConditions requestConditions,
         Duration timeout, Context context) {
+        uploadWithResponse(data, length, parallelTransferOptions, headers, metadata, tier, requestConditions, timeout);
+    }
+
+    /**
+     * Creates a new blob, or updates the content of an existing blob.
+     * <p>
+     * To avoid overwriting, pass "*" to {@link BlobRequestConditions#setIfNoneMatch(String)}.
+     *
+     * @param data The data to write to the blob.
+     * @param length The exact length of the data. It is important that this value match precisely the length of the
+     * data provided in the {@link InputStream}.
+     * @param parallelTransferOptions {@link ParallelTransferOptions} used to configure buffered uploading.
+     * @param headers {@link BlobHttpHeaders}
+     * @param metadata Metadata to associate with the blob.
+     * @param tier {@link AccessTier} for the destination blob.
+     * @param requestConditions {@link BlobRequestConditions}
+     * @param timeout An optional timeout value beyond which a {@link RuntimeException} will be raised.
+     */
+    public void uploadWithResponse(InputStream data, long length, ParallelTransferOptions parallelTransferOptions,
+        BlobHttpHeaders headers, Map<String, String> metadata, AccessTier tier, BlobRequestConditions requestConditions,
+        Duration timeout) {
         final ParallelTransferOptions validatedParallelTransferOptions =
             ModelHelper.populateAndApplyDefaults(parallelTransferOptions);
-        // BlobOutputStream will internally handle the decision for single-shot or multi-part upload.
-        BlobOutputStream blobOutputStream = BlobOutputStream.blockBlobOutputStream(client,
-            validatedParallelTransferOptions, headers, metadata, tier, requestConditions);
-        try {
-            StorageImplUtils.copyToOutputStream(data, length, blobOutputStream);
-            blobOutputStream.close();
-        } catch (IOException e) {
-            Throwable cause = e.getCause();
-            if (cause instanceof BlobStorageException) {
-                throw logger.logExceptionAsError((BlobStorageException) cause);
-            } else {
-                throw logger.logExceptionAsError(new UncheckedIOException(e));
+        Mono<Object> upload = Mono.fromCallable(() -> {
```

Comment:
With this change should we just have this call into `BlockBlobClient`'s upload method?